src/tss2-esys/esys_tr.c

```diff
@@ -133,11 +133,35 @@ Esys_TR_FromTPMPublic_Async(ESYS_CONTEXT * esys_context,
     _ESYS_ASSERT_NON_NULL(esys_context);
     ESYS_TR esys_handle = esys_context->esys_handle_cnt++;
     RSRC_NODE_T *esysHandleNode = NULL;
-    r = esys_CreateResourceObject(esys_context, esys_handle, &esysHandleNode);
-    goto_if_error(r, "Error create resource", error_cleanup);
+    RSRC_NODE_T *node_rsrc = NULL;
+    RSRC_NODE_T *next_node_rsrc;
+
+    for (node_rsrc = esys_context->rsrc_list; node_rsrc != NULL;
+         node_rsrc = next_node_rsrc) {
+         if (node_rsrc->rsrc.handle == tpm_handle) {
+             esysHandleNode = node_rsrc;
+             esys_context->esys_handle = node_rsrc->esys_handle;
+             break;
+         }
+         next_node_rsrc = node_rsrc->next;
+    }
 
-    esysHandleNode->rsrc.handle = tpm_handle;
-    esys_context->esys_handle = esys_handle;
+    if (!esysHandleNode) {
+        /* Object was already created */
+        esys_handle = esys_context->esys_handle_cnt++;
+        r = esys_CreateResourceObject(esys_context, esys_handle, &esysHandleNode);
+        goto_if_error(r, "Error create resource", error_cleanup);
+
+        /* In the first trial no session will be used to determine the object name. */
+        esys_context->sav_session = shandle1;
+        esys_context->session_tab[0] = NULL;
+        esysHandleNode->rsrc.handle = tpm_handle;
+        esys_context->esys_handle = esys_handle;
+        shandle1 = ESYS_TR_NONE;
```

Comment:
So this means that the user specified a session but the object is not in the ESYS_TR list and the call will now use no session handle to go to the TPM to get the name. Shouldn't this error instead as to not surprise the user that the tpm communications were unprotected? What about shandles 2 and 3?

Reply:
Ok lets add a check that if it's not an NV index error. But how is nv index ok? We still don't know the name until we get the public for the attribute and other properties right?

Reply:
**Addition:** The spec says:
"The ‘optionalSession’ parameter can be used in order to pass an HMAC session, encryption
session or audit session. This parameter can also be ESYS_TR_NONE. It is recommended to
pass in one salted HMAC session for untrusted connections to a TPM."
with:
`ESYS_TR optionalSession1, /* optional (ESYS_TR_NONE) */`
which is a contradiction to the "shall" in:
For all other handles (besides NV-Index) this function shall
only be used if the path to the TPM can be trusted.

Reply:
> Yes in all cases it's necessary to determine the name for the HMAC session. So in the case of an NV index it's the users responsibility to provide a session. If the session is provided the public data of the two calls will be compared.

Wait, two calls? Do you mean this function calls readpublic twice or if the user issues 2 Esys_TR_FromTPMPublic calls?

The thing I want to prevent is the case where a user is thinking they're using an HMAC session and that HMAC session getting silently ignored. Despite what the documentation says, we can't silently be dropping sessions.

> For handles of persistent keys BAD_VALUE will be returned if no session is used. So calls of `Esys_TR_FromTPMPublic` for persistent keys which did work in the past now would fail.

Apparently I don't get whats going on here. If we're communicating with the TPM it's because we don't know the name and have to go get it. So how could anything requiring the name, like an HMAC session, work?
The only case I care about is when a user supplies a session and that session is changed to ESYS_TR_NONE.

I really see an API where a user can say make me an esys TR from this TPM handle *and* this name.

> Currently we have one integration test which has to be modified (esys-tr-fromTpmPublic-key.int.c). I also checked the tpm-tools. Here this function is only used with NV indexes. Should I add the test?

Reply:
The ESAPI spec says:
> NOTE: If a session is passed into this function, then the ReadPublic command with the TPM will be called
> twice. The first call is executed without a session in order to retrieve the public name that is then used in
> the HMAC calculation for the second call. For the synchronous invocation the two TPM calls are made
> transparently for the application.
So that's what's implemented in the current PR. If the user is using a HMAC session the session is not ignored.
The second call with the HMAC session will be used to verify the result of the first call.

Reply:
> > transparently for the application.

Ahh ok, but the attacker can still give a bad object, but the path must be trusted in this case. That's fine.

Reply:
Actually it does support serialized EYS_TR's (I thought it did):
```bash
tpm2 createprimary -c primary.ctx
tpm2 evictcontrol -c primary.ctx -o primary.tr
tpm2 create -C primary.tr -u key.pub -r key.priv
```

Reply:
yes the serialization is supported. I just checked the tools to see whether the discussed test of a session for persistent key handles for `Esys_TR_FromTPMPublic`would cause problems. 
But now I would say the check is not necessary because in the parameter description the usage of the session is only recommended. If the check is not needed the would the PR be ok?